Mathlib/RingTheory/PiTensorProduct.lean

```diff
@@ -0,0 +1,262 @@
+  left_distrib := lmul_add
+  right_distrib := add_lmul
+  zero_mul := zero_lmul
+  mul_zero := lmul_zero
+
+instance algebra : Algebra R (⨂[R] i, A i) where
+  __ := hasSMul'
+  toFun := (· • 1)
+  map_one' := by simp
+  map_mul' r s :=show (r * s) • 1 = lmul (r • 1) (s • 1)  by
+    rw [map_smul, map_smul, LinearMap.smul_apply, mul_comm, mul_smul]
+    congr
+    show (1 : ⨂[R] i, A i) = 1 * 1
+    rw [mul_one]
+  map_zero' := by simp
+  map_add' := by simp [add_smul]
+  commutes' r x := by
+    simp only [RingHom.coe_mk, MonoidHom.coe_mk, OneHom.coe_mk]
+    change lmul _ _ = lmul _ _
+    rw [map_smul, map_smul, LinearMap.smul_apply]
+    change r • (1 * x) = r • (x * 1)
+    rw [mul_one, one_mul]
+  smul_def' r x := by
+    simp only [RingHom.coe_mk, MonoidHom.coe_mk, OneHom.coe_mk]
+    change _ = lmul _ _
+    rw [map_smul, LinearMap.smul_apply]
+    change _ = r • (1 * x)
+    rw [one_mul]
+
+/--
+the map `Aᵢ ⟶ ⨂ᵢ Aᵢ` given by `a ↦ 1 ⊗ ... ⊗ a ⊗ 1 ⊗ ...`
+-/
+@[simps]
+def fromComponentAlgHom [DecidableEq ι] (i : ι) : A i →ₐ[R] ⨂[R] i, A i where
+  toFun a := tprod R (Function.update 1 i a)
+  map_one' := by simp only [update_one]; rfl
+  map_mul' a a' := by
+    simp only [tprod_mul_tprod]
+    congr
+    ext j
+    simp only [update, Pi.one_apply, Pi.mul_apply]
+    aesop
+  map_zero' := MultilinearMap.map_update_zero _ _ _
+  map_add' _ _ := MultilinearMap.map_add _ _ _ _ _
+  commutes' r := show tprodCoeff R _ _ = r • tprodCoeff R _ _ by
+    rw [Algebra.algebraMap_eq_smul_one]
+    erw [smul_tprodCoeff]
+    rfl
+
+/--
+Lifting a multilinear map to an algebra homomorphism from tensor product
+-/
+@[simps]
+def liftAlgHom {S : Type*} [Semiring S] [Algebra R S]
+    (f : MultilinearMap R A S)
+    (one : f 1 = 1) (mul : ∀ x y, f (x * y) = f x * f y) : (⨂[R] i, A i) →ₐ[R] S where
+  toFun := lift f
+  map_one' := show lift f (tprod R 1) = 1 by simp [one]
+  map_mul' x y := show lift f (x * y) = lift f x * lift f y by
+    induction' x using PiTensorProduct.induction_on with rx x x₁ x₂ hx₁ hx₂
+    · induction' y using PiTensorProduct.induction_on with ry y y₁ y₂ hy₁ hy₂
+      · simp only [Algebra.mul_smul_comm, Algebra.smul_mul_assoc, tprod_mul_tprod, map_smul,
+          lift.tprod, mul]
+      · simp only [Algebra.smul_mul_assoc, map_smul, lift.tprod, map_add] at hy₁ hy₂ ⊢
+        rw [mul_add, map_add, smul_add, hy₁, hy₂, mul_add, smul_add]
+    · simp only [map_add] at hx₁ hx₂ ⊢
+      rw [add_mul, map_add, hx₁, hx₂, add_mul]
+  map_zero' := by simp only [map_zero]
+  map_add' x y := by simp only [map_add]
+  commutes' r := show lift f (r • tprod R 1) = _ by
+    rw [map_smul, lift.tprod, one, Algebra.algebraMap_eq_smul_one]
+
+end Semiring
+
+noncomputable section Ring
+
+variable [CommRing R] [∀ i, Ring (A i)] [∀ i, Algebra R (A i)]
+
+instance ring : Ring (⨂[R] i, A i) where
+  __ := semiring
+  __ := inferInstanceAs <| AddCommGroup (⨂[R] i, A i)
+
+end Ring
+
+noncomputable section CommRing
```

Comment:
Can you add `CommSemiring` too?

Reply:
(and perhaps even the non-associative/non-unital versions; they shouldn't be much more work, since they just package the same proofs in smaller pieces)

Reply:
I don't think `lmul` works for non-associate/non-unital version because `lmul` really need `Algebra R (A i)` instances so that `x * update z i (r • a) = update (x * z) i (r•(x i * a)))` is true

Reply:
done